cms/server/contest/handlers/delayrequest.py

```
@@ -0,0 +1,109 @@
+from .contest import ContestHandler
+
+
+logger = logging.getLogger(__name__)
+
+
+def N_(msgid):
+    return msgid
+
+
+class DelayRequestHandler(ContestHandler):
+    """Called when the user submits a delay request.
+
+    """
+    @tornado.web.authenticated
+    @multi_contest
+    def post(self):
+        requested_start_time_str = self.get_argument("requested_start_time", "")
+        reason = self.get_argument("reason", "")
+
+        if not requested_start_time_str or not reason:
+            self.notify_error(N_("Invalid request"),
+                            N_("Please provide both a requested start time and a reason."))
+            self.redirect(self.contest_url("communication"))
+            return
+
+        if len(reason) > DelayRequest.MAX_REASON_LENGTH:
+            self.notify_error(N_("Reason too long"),
+                            N_("The reason must be at most %d characters long."),
+                            DelayRequest.MAX_REASON_LENGTH)
+            self.redirect(self.contest_url("communication"))
+            return
+
+        try:
+            naive_dt = datetime.fromisoformat(requested_start_time_str)
+        except (ValueError, TypeError):
+            self.notify_error(N_("Invalid date"),
+                            N_("The requested start time is not valid."))
+            self.redirect(self.contest_url("communication"))
+            return
+
+        tz = get_timezone(self.current_user.user, self.contest)
+        try:
+            if hasattr(tz, "localize"):
+                local_dt = tz.localize(naive_dt, is_dst=None)
+            else:
+                local_dt = naive_dt.replace(tzinfo=tz)
+            
+            utc_dt = local_dt.astimezone(utc)
+            requested_start_time = utc_dt.replace(tzinfo=None)
+        except Exception as e:
+            logger.warning("Timezone conversion failed for delay request: %s. "
+                          "Treating input as UTC.", e)
+            requested_start_time = naive_dt
```

This should raise an error as well; Otherwise we risk the start time of the contest to be not as requested by the user.